### ISSUE TITLE
A segfault when .envrc sources itself recursively

### ISSUE THREAD

**Issue by kaukas** [direnv/direnv#387]
### Steps to reproduce

```bash
$ echo "source $(pwd)/.envrc" > .envrc
$ direnv allow
```

I get `direnv: error signal: segmentation fault` after a while.

I know it is stupid and not direnv's fault... But can anything be done to make the error a bit more self descriptive? I spent a while debugging, logging out and in, and restarting the computer before checking the actual `.envrc` file.

That said, my setup is slightly more complicated (`.envrc` symlinks into a file, and I `mv`ed another file over a symlink which overwrote the actual file...) which is why it took me a while to figure out. I understand if you just close this issue.

## Technical details

* `direnv version`: 2.17.0
* OS release: Ubuntu 18.04

**Comment by zimbatm** [comment]
Interesting corner case :)

One solution I can think of is to introduce a `source_once()` function that keeps records of all previously sourced files and aborts if it already did. It would add a bit of overhead but probably wouldn't be too bad.

**Comment by zimbatm** [comment]
I'm not sure if it would work on macOS because they are stuck on a old GPLv2 version of Bash. For this it would be best to use a bash hash variable to record all the paths

**Comment by kaukas** [comment]
I don't think `source_once` is a good idea; a lot of complexity for a rare edge case. The only improvement I'd expect in this case is perhaps a more informative error message.

If I `echo "source $(pwd)/file.sh" > file.sh && source file.sh` I get `Segmentation fault: 11`. So it seems (and [some internets agree](https://stackoverflow.com/questions/41084393/osx-bash-command-gives-segmentation-fault-11#41086761)) that segfault 11 could happen because of infinite recursion, among other things. If you would not mind improving anything I'd suggest mentioning "perhaps you have infinite recursion?" somewhere in the error message on segfault 11 (if that's at all possible).

Thank you!

**Comment by zimbatm** [comment]
Right, I could probably add a special check if the exit status is 139 instead.

```
$ echo "source ./rec" > rec
$ bash
$ source rec
Segmentation fault
$ echo $?
139
```

It involves drilling into the os-specific sys.Interface in https://golang.org/pkg/os/#ProcessState since exit statuses are apparently not something Windows supports.

I will wait to see if anyone else is hitting that issue before going further.

**Comment by Sillium** [comment]
I get this too with direnv 2.20.1.

I have:
- directory structure a/b/c with a .envrc in each of the three directories
- a ". ../.envrc" as the first line in a/b/.envrc and in a/b/c/.envrc

I get
`direnv allow .
direnv: loading .envrc                                                                                              
direnv: error signal: segmentation fault`
when executing `direnv allow .` in directory c.
